services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java

```diff
@@ -960,19 +956,31 @@ public static Response redirectAfterSuccessfulFlow(KeycloakSession session, Real
 
     }
 
-    public static String getSessionIdFromSessionCookie(KeycloakSession session) {
+    /**
+     * @param session keycloak session
+     * @param sessionId in plain-text
+     * @return true if sessionId matches with the session from KEYCLOAK_SESSION_COOKIE
+     */
+    public static boolean compareSessionIdWithSessionCookie(KeycloakSession session, String sessionId) {
+        if (sessionId == null) {
+            throw new IllegalArgumentException("Not expected to provide null sessionId");
+        }
+
         String cookie = session.getProvider(CookieProvider.class).get(CookieType.SESSION);
         if (cookie == null || cookie.isEmpty()) {
             logger.debugv("Could not find cookie: {0}", KEYCLOAK_SESSION_COOKIE);
-            return null;
+            return false;
         }
 
-        String[] parts = cookie.split("/", 3);
-        if (parts.length != 3) {
-            logger.debugv("Cannot parse session value from: {0}", KEYCLOAK_SESSION_COOKIE);
-            return null;
+        if (cookie.equals(sha256UrlEncodedHash(sessionId))) return true;
+
+        // Backwards compatibility
```

Comment:
Should we log an issue to remove this in the next major?

Reply:
Logged https://github.com/keycloak/keycloak/issues/34772